File: backend/model/reservation.go
package model

import (
	"database/sql"
	"time"
)

type ReservationAPI struct {
	CreatorID      string    `json:"creator_id"`
	StartAt        time.Time `json:"start_at"`
	EndAt          time.Time `json:"end_at"`
	PlanID         int64     `json:"plan_id"`
	UpdatePlanMemo bool      `json:"update_plan_memo"`
	PlanMemo       string    `json:"plan_memo"`
	UserIDs        []string  `json:"user_ids"`
	FacilityIDs    []int64   `json:"facility_ids"`
}

type Reservation struct {
	ID        int64       `json:"id"`
	Creator   *User       `json:"creator"`
	StartAt   time.Time   `json:"start_at"`
	EndAt     time.Time   `json:"end_at"`
	Plan      *Plan       `json:"plan"`
	PlanMemo  string      `json:"plan_memo"`
	CreatedAt time.Time   `json:"created_at"`
	UpdatedAt time.Time   `json:"updated_at"`
	Attendees []*User     `json:"attendees"`
	Places    []*Facility `json:"places"`
}

type ReservationDTO struct {
	ID        int64     `db:"id"`
	CreatorID string    `db:"creator_id"`
	StartAt   time.Time `db:"start_at"`
	EndAt     time.Time `db:"end_at"`
	PlanID    int64     `db:"plan_id"`
	PlanMemo  string    `db:"plan_memo"`
	CreatedAt time.Time `db:"created_at"`
	UpdatedAt time.Time `db:"updated_at"`
}

type ReservationDTOWithStruct struct {
	ID         int64          `db:"id"`
	CreatorID  string         `db:"creator_id"`
	StartAt    time.Time      `db:"start_at"`
	EndAt      time.Time      `db:"end_at"`
	PlanID     int64          `db:"plan_id"`
	PlanMemo   string         `db:"plan_memo"`
	CreatedAt  time.Time      `db:"created_at"`
	UpdatedAt  time.Time      `db:"updated_at"`
	UserID     sql.NullString `db:"user_id"`
	FacilityID sql.NullInt64  `db:"facility_id"`
}

type RangeAPI struct {
	From string `form:"from"`
	To   string `form:"to"`
}

type DateRange struct {
	From time.Time
	To   time.Time
}

File: frontend/src/api/reservation.ts
import { baseInstance } from '.'
import { Reservation, ReservationDTO, convertToReservation, DateRange, CodeResponse } from '@types'

export const getAllReservations = async (): Promise<Reservation[]> => {
  return baseInstance.get<ReservationDTO[]>('/reservation').then(res => {
    if (res.status === 200) {
      return res.data.map(reservation => convertToReservation(reservation))
    }
    return null
  }).catch(() => {
    return null
  })
}

export const getAllReservationsInRange = async (r: DateRange): Promise<Reservation[]> => {
  return baseInstance.get<ReservationDTO[]>('/reservation', {
    params: {
      from: r.from,
      to: r.to,
    },
  }).then(res => {
    if (res.status === 200) {
      return res.data.map(reservation => convertToReservation(reservation))
    }
    return null
  }).catch(() => {
    return null
  })
}

export const getReservation = async (id: string): Promise<Reservation> => {
  return baseInstance.get<ReservationDTO>(`/reservation/${id}`).then(res => {
    if (res.status === 200) {
      return convertToReservation(res.data)
    }
    return null
  }).catch(() => {
    return null
  })
}

export const createReservation = async (r: Reservation): Promise<Reservation> => {
  return baseInstance.post<ReservationDTO>('/reservation', {
    creator_id: r.creator.id,
    start_at: r.startAt.format(),
    end_at: r.endAt.format(),
    plan_id: r.plan.id,
    plan_memo: r.planMemo,
    user_ids: r.attendees.map(attendee => attendee.id),
    facility_ids: r.places.map(place => place.id)
  }).then(res => {
    if (res.status === 200) {
      return convertToReservation(res.data)
    }
    return null
  }).catch(() =>{
    return null
  })
}

export const updateReservation = async (r: Reservation): Promise<Reservation> => {
  return baseInstance.put<ReservationDTO>(`/reservation/${r.id}`, {
    creator_id: r.creator.id,
    start_at: r.startAt.format(),
    end_at: r.endAt.format(),
    plan_id: r.plan.id,
    plan_memo: r.planMemo,
    user_ids: r.attendees.map(attendee => attendee.id),
    facility_ids: r.places.map(place => place.id)
  }).then(res => {
    if (res.status === 200) {
      return convertToReservation(res.data)
    }
    return null
  }).catch(() =>{
    return null
  })
}

export const deleteReservation = async (r: Reservation): Promise<CodeResponse> => {
  return baseInstance.delete<CodeResponse>(`/reservation/${r.id}`).then(res => {
    if (res.status === 204) {
      return {
        code: 204,
        message: "deleted",
      }
    }
    return null
  }).catch(() =>{
    return null
  })
}
